### PR TITLE
fix: [DEL-5715]: Detect mismatch between the delegate yaml and the delegate image and display error message 

### PR DESCRIPTION
**Pull request by raghuAtWings**
## Describe your changes
Fail startup of an immutable delegate if customer used legacy delegate yaml to start it
Screenshot 2023-01-11 at 3 08 42 PM

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/42405)

### CODE REVIEW

**Review by mdmilic — CHANGES_REQUESTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42405)* status: 1 file(s) to review. 1 change requests. 0 of 1 approvals obtained. 

___
*[`260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java` line 465 at r1](https://reviewable.io/reviews/harness/harness-core/42405#-NLN7-x058LJzASr3fnl:-NLN7-x058LJzASr3fnm:b-lsaywx) ([raw file](https://github.com/harness/harness-core/blob/da7a5119925a2d2f70866725107ccbeb321a346b/260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java#L465)):*
> ```Java
>    * This will indicate that the customer started an immutable delegate with older stateful set yaml
>    */
>   private void checkForImmutbleAndStatefulset() {
> ```

You should limit this only for K8S delegate type (& Helm), for docker and shell hostname can be anything. 
___
*[`260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java` line 476 at r1](https://reviewable.io/reviews/harness/harness-core/42405#-NLN7HO165Id6RrGMgwo:-NLN7HO165Id6RrGMgwp:byo0m1v) ([raw file](https://github.com/harness/harness-core/blob/da7a5119925a2d2f70866725107ccbeb321a346b/260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java#L476)):*
> ```Java
> 
>     try {
>       Integer.parseInt(HOST_NAME.substring(index));
> ```

Technically the last part of immutable hostname can also be a valid integer. It is supposed to be a 5-digit hash value, but in extreme case that can also contain only numbers, in which case this would fail. Maybe if parsed correctly check if it's less than some big number (e.g. 100). For stateful set this number represents the index of a replica which always start at 0 AFAIK. For Deployment it's it would have to be 5 digits. 
Otherwise the format for each is below, but it would be hard to parse I think (because Resource name can be anything and doesn't need to match delegate name):
<resource_name>-<pod-index> <-- might have extra stuff for multi container pods
<resource_name>-<deployment_hash>-<pod-hash>

**Review by mdmilic — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42405)* status: 1 file(s) to review. 1 change requests. 0 of 1 approvals obtained. 

___
*[`260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java` line 476 at r1](https://reviewable.io/reviews/harness/harness-core/42405#-NLN7HO165Id6RrGMgwo:-NLN9JHL--B35HTF2Ok2:b-56ltmk) ([raw file](https://github.com/harness/harness-core/blob/da7a5119925a2d2f70866725107ccbeb321a346b/260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java#L476)):*
<details><summary><i>Previously, mdmilic (Marko Milic) wrote…</i></summary><blockquote>

Technically the last part of immutable hostname can also be a valid integer. It is supposed to be a 5-digit hash value, but in extreme case that can also contain only numbers, in which case this would fail. Maybe if parsed correctly check if it's less than some big number (e.g. 100). For stateful set this number represents the index of a replica which always start at 0 AFAIK. For Deployment it's it would have to be 5 digits. 
Otherwise the format for each is below, but it would be hard to parse I think (because Resource name can be anything and doesn't need to match delegate name):
<resource_name>-<pod-index> <-- might have extra stuff for multi container pods
<resource_name>-<deployment_hash>-<pod-hash>

</blockquote></details>

parts of the format were cut of by GH so here it goes again:
<resource_name> - <pod_index>
<resource_name> - <deployment_hash> - <pod_hash>

**Comment by raghuAtWings** [comment]
*[`260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java` line 476 at r1](https://reviewable.io/reviews/harness/harness-core/42405#-NLN7HO165Id6RrGMgwo:-NLNo9Uy1ijhozXcsGdP:bx4jkzw) ([raw file](https://github.com/harness/harness-core/blob/29267b9d919b8a09db8d8d5df405b573bccbdb0b/260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java#L476)):*
<details><summary><i>Previously, mdmilic (Marko Milic) wrote…</i></summary><blockquote>

parts of the format were cut of by GH so here it goes again:
<resource_name> - <pod_index>
<resource_name> - <deployment_hash> - <pod_hash>
</blockquote></details>

i am looking for index of "-" and then checking if the remaining part is an integer

**Comment by raghuAtWings** [comment]
*[`260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java` line 465 at r1](https://reviewable.io/reviews/harness/harness-core/42405#-NLN7-x058LJzASr3fnl:-NLNoSra0ZYIkXKPPBBm:bb7f2mf) ([raw file](https://github.com/harness/harness-core/blob/29267b9d919b8a09db8d8d5df405b573bccbdb0b/260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java#L465)):*
<details><summary><i>Previously, mdmilic (Marko Milic) wrote…</i></summary><blockquote>

You should limit this only for K8S delegate type (& Helm), for docker and shell hostname can be anything. 
</blockquote></details>

done

**Review by mdmilic — CHANGES_REQUESTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42405)* status: 1 file(s) to review. 1 change requests. 0 of 1 approvals obtained. 

___
*[`260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java` line 476 at r1](https://reviewable.io/reviews/harness/harness-core/42405#-NLN7HO165Id6RrGMgwo:-NLNqH0-BkvVS03m0XLB:b8pt2h5) ([raw file](https://github.com/harness/harness-core/blob/29267b9d919b8a09db8d8d5df405b573bccbdb0b/260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java#L476)):*
<details><summary><i>Previously, raghuAtWings (Raghvendra Singh) wrote…</i></summary><blockquote>

i am looking for index of "-" and then checking if the remaining part is an integer
</blockquote></details>

I understand, but you are using lastIndexOf. This is example of a valid immutable delegate hostname:
`markom-cg-imm-dfzqqy-8bbd86b7b-23455`
HOST_NAME.substring(index) would be `23455` which is a valid integer, so you would kill the delegate even though it's an immutable delegate on a Deployment resource type.

**Review by Vbhatia20 — DISMISSED** [review]
This change is intended for K8 and HELM delegate.

**Review by gaurav-jain1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42405)* status: 1 file(s) to review. 1 change requests. 0 of 1 approvals obtained. 

___
*[`260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java` line 482 at r3](https://reviewable.io/reviews/harness/harness-core/42405#-NLPEHuj6N235cQBRu73:-NLPEHuj6N235cQBRu74:bjsljfx) ([raw file](https://github.com/harness/harness-core/blob/e3d8849d076f908807586ae21720354cd66576c3/260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java#L482)):*
> ```Java
>         log.error("It appears that you have used a legacy delegate yaml with the newer delegate image."
>             + " Please note that for the delegate images formatted as YY.MM.XXXXX you should download a fresh yaml and not reuse legacy delegate yaml");
>         System.exit(1);
> ```

nit: instead we can call `initiateSelfDestruct()`, so that pod itself can destroy.



_Code quote:_
```Java
System.exit(1);
```

**Review by xingchi-jin — DISMISSED** [review]
This PR can solve the problem. The approach of checking the last number of hostname makes sense 
The comments above from @mdmilic  and @gaurav-jain1 make sense

I will approve it. But I want to further check if we could do this in start up probe in the future. benefit is, it's easier to remove/edit when customer accidentally block on it. The error message is also more obviously surfaced on k8s, users don't need to go to logs 

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42405#-:-NLSPf1zAsF_Q6T1K3lC:b1obzx3)* status: 1 file(s) to review. 1 change requests.

**Comment by raghuAtWings** [comment]
trigger ti2

**Comment by raghuAtWings** [comment]
trigger ti0

**Comment by raghuAtWings** [comment]
trigger tin0

**Comment by raghuAtWings** [comment]
trigger ti1

**Comment by raghuAtWings** [comment]
trigger ti3

**Review by raghuAtWings — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42405)* status: 1 file(s) to review. 1 change requests. 1. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java` line 476 at r1](https://reviewable.io/reviews/harness/harness-core/42405#-NLN7HO165Id6RrGMgwo:-NLXeKUU4kvZJ1liVmVF:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/29267b9d919b8a09db8d8d5df405b573bccbdb0b/260-delegate/src/main/java/io/harness/delegate/service/DelegateAgentServiceImpl.java#L476)):*
<details><summary><i>Previously, mdmilic (Marko Milic) wrote…</i></summary><blockquote>

I understand, but you are using lastIndexOf. This is example of a valid immutable delegate hostname:
`markom-cg-imm-dfzqqy-8bbd86b7b-23455`
HOST_NAME.substring(index) would be `23455` which is a valid integer, so you would kill the delegate even though it's an immutable delegate on a Deployment resource type. 
</blockquote></details>

Done.

**Review by xingchi-jin — APPROVED** [review]
> Reviewed 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/42405)* status: 1 change requests. 1. unresolved discussions.
